ovirt_imageio/client/_ui.py

```diff
@@ -101,24 +109,41 @@ def close(self):
 
     def _draw(self, last=False):
         line = self._format_line()
+        if self._format and not last:
+            line += ','
 
         # Using "\r" moves the cursor to the first column, so the next progress
-        # will overwrite this one. If this is the last progress, we use "\n" to
-        # move to the next line. Otherwise, the next shell prompt will include
-        # part of the old progress.
-        end = "\n" if last else "\r"
+        # will overwrite this one. If this is the last progress or there is a
+        # non-default format set, we use "\n" to move to the next line.
+        # Otherwise, the next shell prompt will include part of the old
+        # progress.
+        end = "\n" if last or self._format else "\r"
 
         self._output.write(line + end)
         self._output.flush()
 
     def _format_line(self):
+        progress_value = max(0, self._value)
         elapsed = self._now() - self._start
-        progress = f"{max(0, self._value):3d}%" if self._size else "----"
-        done = util.humansize(self._done)
-        rate = util.humansize((self._done / elapsed) if elapsed else 0)
-        phase = f" | {self._phase}" if self._phase else ""
-        line = f"[ {progress} ] {done}, {elapsed:.2f} s, {rate}/s{phase}"
-        return line.ljust(self._width, " ")
+        phase = f"{self._phase}" if self._phase else ""
+
+        if self._format == FORMAT_JSON:
+            line = json.dumps({
+                'progress': progress_value,
+                'description': phase,
+                'done': self._done,
+                'elapsed': round(elapsed, 2),
+                })
+        else:
+            progress = f"{progress_value:3d}%" if self._size else "----"
+            done = util.humansize(self._done)
+            rate = util.humansize((self._done / elapsed) if elapsed else 0)
+            if self._phase:
+                phase = f" | {phase}"
+            line = f"[ {progress} ] {done}, {elapsed:.2f} s, {rate}/s{phase}"
+            line = line.ljust(self._width, " ")
```

Comment:
Using TextWriter and JsonWriter (or better name) will make this code much cleaner.